Update Nepali Layouts

sapradhan opened the pull request:
Summary of Changes: 
- Modify Nepali Traditional layout to make it more practical - put high frequency characters ahead of rarely used combinations.
- Retain existing traditional layout as Nepali Traditional MPP, organized comments.
- Use actual Devanagari characters in Romanized layout to improve readability.


-----

Hello, I want to contribute a few changes to Nepali layouts. I have tested these layouts on mobile and tablet. This is my first attempt authoring layouts for Futo keyboard, please let me know if I have missed anything.

Nepali is my native language and my work involves typing in Nepali on a daily basis on desktop as well as on mobile. I have also contributed Nepali layouts to m17n and Keyman. Of all the tools I have worked with, Futo has been the most intuitive and has the best developer experience. Thank you for developing this amazing layout format.

abb128 commented:
Thanks for the PR, I see there are significant changes to Nepali Traditional. In general we want to avoid significantly changing existing layouts beyond fixing mistakes, bugs or obvious usability issues, because it would be unexpected and disorienting for users to update the app and their keyboard layout changes significantly. Do you think it would make sense to add it as a new layout instead?

sapradhan commented:
Thank you for taking time to review. I understand your concern about breaking compatibility with the original layout. However, please let me present my case that, this layout in current form is extremely difficult to use. 

I believe this layout has been ported from android Nepali traditional layout, which is in turn based on [Nepali Traditional desktop layout](https://ltk.org.np/keyboard_layouts.php) by Madan Puraskar Pratisthan, now LTK. The desktop layout has 4 rows while mobile layout has 3, merging numeric and top row, dropping few characters. But strange choices were made while choosing characters to retain, specially in the shift layer. 

Here is comparative analysis of frequencies of characters you get normally vs the ones you get with a longpress on the top row of shifted layer.

| Normal                 |Frequency | Longpress  | Frequency |
|------------------------|----------|------------|-----------|
| त्त (\u0924\u094d\u0924) | 6,293    | ञ (\u091e) |  7,755    |
| ड्ढ (\u0921\u094d\u0922) | 23       | ई (\u0908) |  38,141‍   |
| ऐ (\u0910)             | 1,999    | घ (\u0918) |  17‌,302   |
| ट्ट (\u091f\u094d\u091f) | 1,799    | छ (\u091b) |  74,928   |
| ठ्ठ (\u0920\u094d\u0920) | 100      | ट (\u091f) |  78,342   |
| ऊ (\u090a)             | 761      | ठ (\u0920) |  18,039   |
| क्ष (\u0915\u094d\u0937) | 23,620   | ड (\u0921) |  39,763   |
| इ (\u0907)             | 24,738   | ढ (\u0922) |  10,728   |
| ए (\u090f)             | 81,699   | ण (\u0923) |  41,072   |

*This frequency analysis was performed on 100k Nepali sentences. [source](https://github.com/sapradhan/nepali-char-ngrams)*

As you can see very frequently used characters are placed in the long-press menu, while rarely used [conjuncts](https://en.wikipedia.org/wiki/Devanagari_conjuncts) are assigned to normal keypress. It seems the intent was to cram in as much conjuncts as possible not considering if that would be useful. It is actually more intuitive to type conjuncts with 2-3 presses of constituent keys than to have dedicated keys.  

Changes -
- Shift layer has significant changes 
    - moved frequently used characters to normal 
    - removed several really low frequency conjunct combinations. they can be composed from characters already present in layout
![original_shifted](https://github.com/user-attachments/assets/f606927b-322b-4485-b024-f184918bc685) *Original*
![proposed_shifted](https://github.com/user-attachments/assets/ae0f93c1-6b6d-4273-abb4-3e8ad4edde2b) *Proposed*
- Normal layer is mostly same except for -
    - some characters have been moved so that the mobile layout is more closer to the reference desktop layout. 
    - missing character ऱ has been added in a longpress menu. 
![original_normal](https://github.com/user-attachments/assets/a304e41d-efc2-4faf-a80f-2d877107b797) *Original*
![proposed_normal](https://github.com/user-attachments/assets/b3b6d2b1-1ee4-455e-813a-bb039d5bfe8b) *Proposed*


I created and maintain [Nepali Traditional layout in keyman](https://keyman.com/keyboards/nepali_traditional?bcp47=ne) in similar spirit; to overcome shortcomings of the original android layout.   

The original layout has been retained as Nepali Traditional (MPP) for historical purposes.

I am aware I may be asking you to make a decision about a script that you may not have much familiarity with. And if you feel its a decision you'd rather not make, I am okay to submit this layout in another name (Nepali Traditional Simplified). But given the evidence, I would request to reconsider and not perpetuate the mistakes in the original layout and accept this as the "default" Nepali traditional layout.

Best

abb128 commented on Devanagari/nepali_romanized.yaml:
```diff
@@ -1,128 +1,39 @@
-    # U+0907: इ DEVANAGARI LETTER I
-    # U+0908: ई DEVANAGARI LETTER II
-    - {type: case, normal: ["\u0907"], shifted: ["\u0908"]}
+    - {type: case, normal: ["ट", "॰"], shifted: ["ठ"]}  #Q
```
In the original version this was [TTA, "%", NUKTA]. The % symbol isn't a literal % but a hint to put the number key in that position if it's required according to user settings, otherwise it gets put at the end by default. If it's more common to long-press to type the number rather than that extra key it's recommended to put % so that the number is the default long-press key.

I'm not familiar with this language and I see you also replaced nukta with abbreviation sign so it's your judgment if the number should have priority or if the abbreviation sign should have priority when long-pressing this key

abb128 commented on Devanagari/nepali_romanized.yaml:
```diff
@@ -1,128 +1,39 @@
+    - {type: case, normal: ["ट", "॰"], shifted: ["ठ"]}  #Q
+    - {type: case, normal: ["ौ"], shifted: ["औ"]}  #W
+    - {type: case, normal: ["े"], shifted: ["ै"]}  #E
+    - {type: case, normal: ["र", "र्‍|ऱ्"], shifted: ["ृ", "ॄ"]}  #R
```
See if % is needed here too

sapradhan replied:
thank you for pointing this out. I have added % on both locations and on traditional layout

abb128 commented:
Thank you for the information. If it is a significant improvement I'm happy to accept it as the default traditional, but I also generally try to avoid significantly changing layouts for any potential existing users in case they have muscle memory trained and they can't figure out how to revert to the old one.

IMO if it's worth preserving the old layout for historical purposes, it might be better to just keep it and make the new version an entirely new layout instead (nepali_traditional_v2 or similar). You can edit the name/description fields to indicate the old one is legacy or not recommended, and you can put the new one above the old one in mapping.yaml so that it appears higher in the layout picker. Does that make sense?

I left a comment on the romanized layout with some information about % sign you might want to consider

sapradhan commented:
New layout appearing before legacy is sufficient. I have also added hint in the description of the legacy layout.  Thank you for your guidance.
